From 3ead64173dadf6b72e6e40c2923126c3f5d8273d Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 28 Jun 2021 16:02:26 +0300
Subject: [PATCH] [SW-47071] gaudi: sec: protecting mmu_up block regs

Using protection bits to block access to chosen registers inside
MMU_UP block.

Change-Id: I876affe96b9816cea9d0508413970b63e41bd62e
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi/security.c b/arch/gaudi/security.c
index 4a21eac..8678c37 100644
--- a/arch/gaudi/security.c
+++ b/arch/gaudi/security.c
@@ -32,6 +32,7 @@
 #include <dma_if_w_s_regs.h>
 #include <sync_mngr_w_s_regs.h>
 #include <gic_regs.h>
+#include <mmu_up_regs.h>
 
 #include <cpu_if_masks.h>
 #include <psoc_global_conf_masks.h>
@@ -846,13 +847,36 @@
 	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
 }
 
+void preboot_secure_mmu_up(void)
+{
+	u32 mask, prot_base, base = mmMMU_UP_BASE & ~HL_CFG_BASE;
+
+
+	/* blocking regs within offsets: 0 - 0x80 */
+	prot_base = base;
+	mask = (PROT_BIT(mmMMU_UP_FEATURE_ENABLE) |
+		PROT_BIT(mmMMU_UP_VA_ORDERING_MASK_31_7) |
+		PROT_BIT(mmMMU_UP_VA_ORDERING_MASK_31_7) |
+		PROT_BIT(mmMMU_UP_LOG2_DDR_SIZE) |
+		PROT_BIT(mmMMU_UP_SCRAMBLER) |
+		PROT_BIT(mmMMU_UP_MEM_INIT_BUSY) |
+		PROT_BIT(mmMMU_UP_DBG_MEM_WRAP_RM) |
+		PROT_BIT(mmMMU_UP_PIPE_CREDIT) |
+		PROT_BIT(mmMMU_UP_SLICE_CREDIT) |
+		PROT_BIT(mmMMU_UP_MMU_BYPASS));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	prot_base += SNGL_REG_PROT_SIZE;	/* offsets: 0x80 - 0xF80 */
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
 static void preboot_secure_pci(void)
 {
 	/* Protect PCIE_WRAP - 4K */
 	preboot_secure_pci_wrap();
 
-	/* protect MMU UP priv and conf bits */
-	set_prot_bits_priv_int_conf(mmMMU_UP_BASE);
+	/* protect registers in MMU_UP */
+	preboot_secure_mmu_up();
 
 	/* protect STLB priv and conf bits */
 	set_prot_bits_priv_int_conf(mmSTLB_BASE);
